### PR TITLE
Make concurrency control controllable

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add setting `use_concurrency_control` for better testing of the new concurrency control feature.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 6fddb5bad35073235cf5f92d2e7644cf3ad693fd with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49618/6fddb5bad35073235cf5f92d2e7644cf3ad693fd/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by alexey-milovidov** [comment]
This is for #46927

**Comment by serxa** [comment]
So even with always `use_concurrency_control = false` we have the same error in tests `00913_many_threads`, `00096_aggregation_min_if` and `02514_null_dictionary_source`:
```
2023-08-07 17:30:17 [ip-172-31-92-46] 2023.08.07 17:30:17.124144 [ 168380 ] {b9ad2d5e-ddbd-46db-bf60-dde161542295} <Error> executeQuery: Code: 49. DB::Exception: Can't push thread because it is already in threads queue. (LOGICAL_ERROR) (version 23.8.1.1) (from [::ffff:127.0.0.1]:52948) (comment: 00913_many_threads.sql) (in query: SELECT count() FROM t;), Stack trace (when copying this message, always include the lines below):
2023-08-07 17:30:17 
2023-08-07 17:30:17 0. /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/exception:134: Poco::Exception::Exception(String const&, int) @ 0x000000000e252b92 in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 1. /fasttest-workspace/build/./src/Common/Exception.cpp:98: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000007cf6e77 in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 2. DB::Exception::Exception<char const (&) [57]>(int, char const (&) [57]) @ 0x0000000004473978 in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 3. /fasttest-workspace/build/./src/Processors/Executors/ThreadsQueue.h:39: DB::ExecutorTasks::tryGetTask(DB::ExecutionThreadContext&) @ 0x000000000d8a2b3b in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 4. /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/atomic:958: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x000000000d89b806 in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 5. /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:833: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000000d89c5f9 in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 6. /fasttest-workspace/build/./base/base/../base/wide_integer_impl.h:809: ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x0000000007db6f58 in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 7. /fasttest-workspace/build/./src/Common/ThreadPool.cpp:0: ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() @ 0x0000000007db9fdf in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 8. /fasttest-workspace/build/./base/base/../base/wide_integer_impl.h:809: ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x0000000007db4642 in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 9. /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x0000000007db84ae in /fasttest-workspace/build/programs/clickhouse
2023-08-07 17:30:17 10. ? @ 0x00007fd72bcb5b43 in ?
2023-08-07 17:30:17 11. ? @ 0x00007fd72bd46bb4 in ?
2023-08-07 17:30:17 
2023-08-07 17:30:17 Received exception from server (version 23.8.1):
2023-08-07 17:30:17 Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Can't push thread because it is already in threads queue. (LOGICAL_ERROR)
```

File: src/Processors/Executors/PipelineExecutor.cpp [hunk 1]
```diff
@@ -320,7 +318,8 @@ void PipelineExecutor::spawnThreads()
 {
     while (auto slot = slots->tryAcquire())
     {
-        size_t thread_num = threads++;
+        size_t thread_num = threads;
+        ++threads;
```
Comment (serxa) [inline src/Processors/Executors/PipelineExecutor.cpp:322]: It looks like the race is here, because `threads` is `std::atomic_size_t`

**Comment by serxa** [comment]
Let's now check if CI agrees that it was the only problem and then enable `use_concurrency_control = true` back

**Pull request merged**